File: AINovalServer/src/main/java/com/ainovel/server/domain/model/NovelSettingItem.java
package com.ainovel.server.domain.model;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Map;

import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.index.CompoundIndex;
import org.springframework.data.mongodb.core.index.CompoundIndexes;
import org.springframework.data.mongodb.core.mapping.Document;

import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.experimental.SuperBuilder;

/**
 * 小说设定条目实体
 * 用于存储小说的设定信息，如世界观、人物、地点、物品、纪年史等
 */
@Data
@SuperBuilder
@NoArgsConstructor
@AllArgsConstructor
@Document(collection = "novel_setting_items")
@CompoundIndexes({
    @CompoundIndex(name = "novel_type_name_idx", def = "{'novelId': 1, 'type': 1, 'name': 1}"),
    @CompoundIndex(name = "novel_scene_idx", def = "{'novelId': 1, 'sceneIds': 1}"),
    @CompoundIndex(name = "novel_status_idx", def = "{'novelId': 1, 'status': 1}"),
    @CompoundIndex(name = "novel_parent_idx", def = "{'novelId': 1, 'parentId': 1}")
})
public class NovelSettingItem {
    
    @Id
    private String id;
    
    // 关联的小说ID
    private String novelId;
    
    // 关联的用户ID
    private String userId;
    
    // 设定条目名称
    private String name;
    
    // 设定条目类型 (如：人物、地点、物品、时间线等)
    private String type;
    
    // 设定条目描述
    private String description;
    
    // 设定条目属性 (键值对形式存储各种属性)
    private Map<String, String> attributes;
    
    // 设定条目图像URL (如有)
    private String imageUrl;
    
    // 与其他设定条目的关系
    private List<SettingRelationship> relationships;
    
    // 关联的场景ID列表
    private List<String> sceneIds;
    
    // 设定条目优先级 (1-10，用于控制在相关性相似时的排序)
    private Integer priority;
    
    // 生成方式 (manual, ai_generated, imported, AI_SETTING_GENERATION)
    private String generatedBy;
    
    // 设定条目标签
    private List<String> tags;
    
    // 设定条目状态 (active, inactive, draft, SUGGESTED)
    private String status;
    
    // 设定向量数据 (存储嵌入后的向量，用于相似性搜索)
    private List<Float> vector;
    
    // 创建时间
    private LocalDateTime createdAt;
    
    // 最后更新时间
    private LocalDateTime updatedAt;
    
    // 该设定条目是否为待审核的AI建议
    private boolean isAiSuggestion;
    
    // 额外元数据
    private Map<String, Object> metadata;
    
    // ==================== 父子关系字段 ====================
    
    // 父设定ID (建立层级关系的核心字段)
    private String parentId;
    
    // 子设定ID列表 (冗余字段，用于快速查询)
    private List<String> childrenIds;
    
    // ==================== AI上下文追踪字段 ====================
    
    // 名称/别名追踪设置 (track, no_track)
    private String nameAliasTracking;
    
    // AI上下文包含设置 (always, detected, dont_include, never)
    private String aiContextTracking;
    
    // 设定引用更新设置 (update, ask, no_update)
    private String referenceUpdatePolicy;
    
    /**
     * 设定关系实体
     * 描述设定条目之间的关系
     */
    @Data
    @lombok.Builder
    @NoArgsConstructor
    @AllArgsConstructor
    public static class SettingRelationship {
        
        // 目标设定条目ID
        private String targetItemId;
        
        // 关系类型 (如：父子关系、友谊关系、敌对关系、地理包含等)
        private String type;
        
        // 关系描述
        private String description;
        
        // 关系强度 (1-10)
        private Integer strength;
        
        // 关系方向 (单向、双向)
        private String direction;
        
        // 关系创建时间
        private LocalDateTime createdAt;
        
        // 关系额外属性
        private Map<String, Object> attributes;
    }
} 

File: AINoval/lib/models/novel_setting_item.dart
import 'dart:convert';
import 'package:equatable/equatable.dart';
import 'package:ainoval/models/ai_context_tracking.dart';
import 'package:ainoval/models/setting_relationship_type.dart';

/// 小说设定条目模型
class NovelSettingItem extends Equatable {
  final String? id;
  final String? novelId;
  final String? userId;
  final String name;
  final String? type;
  final String? content;
  final String? description;
  final Map<String, String>? attributes;
  final String? imageUrl;
  final List<SettingRelationship>? relationships;
  final List<String>? sceneIds;
  final int? priority;
  final String? generatedBy;
  final List<String>? tags;
  final String? status;
  final List<double>? vector;
  final DateTime? createdAt;
  final DateTime? updatedAt;
  final bool isAiSuggestion;
  final Map<String, dynamic>? metadata;
  
  // ==================== 父子关系字段 ====================
  
  /// 父设定ID（建立层级关系的核心字段）
  final String? parentId;
  
  /// 子设定ID列表（冗余字段，用于快速查询）
  final List<String>? childrenIds;
  
  // ==================== AI上下文追踪字段 ====================
  
  /// 名称/别名追踪设置
  final NameAliasTracking nameAliasTracking;
  
  /// AI上下文包含设置
  final AIContextTracking aiContextTracking;
  
  /// 设定引用更新设置
  final SettingReferenceUpdate referenceUpdatePolicy;

  const NovelSettingItem({
    this.id,
    this.novelId,
    this.userId,
    required this.name,
    this.type,
    this.content = "",
    this.description,
    this.attributes,
    this.imageUrl,
    this.relationships,
    this.sceneIds,
    this.priority,
    this.generatedBy,
    this.tags,
    this.status,
    this.vector,
    this.createdAt,
    this.updatedAt,
    this.isAiSuggestion = false,
    this.metadata,
    this.parentId,
    this.childrenIds,
    this.nameAliasTracking = NameAliasTracking.track,
    this.aiContextTracking = AIContextTracking.detected,
    this.referenceUpdatePolicy = SettingReferenceUpdate.ask,
  });

  factory NovelSettingItem.fromJson(Map<String, dynamic> json) {
    List<SettingRelationship>? relationships;
    if (json['relationships'] != null && json['relationships'] is List) {
      relationships = (json['relationships'] as List)
          .map((e) => SettingRelationship.fromJson(e as Map<String, dynamic>))
          .toList();
    }

    Map<String, String>? attributesMap;
    if (json['attributes'] != null && json['attributes'] is Map) {
      attributesMap = Map<String, String>.from(json['attributes'] as Map);
    }

    List<String>? tagsList;
    if (json['tags'] != null && json['tags'] is List) {
      tagsList = List<String>.from(json['tags'] as List);
    }
    
    List<String>? sceneIdsList;
    if (json['sceneIds'] != null && json['sceneIds'] is List) {
      sceneIdsList = List<String>.from(json['sceneIds'] as List);
    }
    
    List<String>? childrenIdsList;
    if (json['childrenIds'] != null && json['childrenIds'] is List) {
      childrenIdsList = List<String>.from(json['childrenIds'] as List);
    }

    List<double>? vectorList;
    if (json['vector'] != null && json['vector'] is List) {
      vectorList = (json['vector'] as List).map((e) => (e as num).toDouble()).toList();
    }
    
    Map<String, dynamic>? metadataMap;
    if (json['metadata'] != null && json['metadata'] is Map) {
      metadataMap = Map<String, dynamic>.from(json['metadata'] as Map);
    }

    return NovelSettingItem(
      id: json['id'] as String?,
      novelId: json['novelId'] as String?,
      userId: json['userId'] as String?,
      name: json['name'] as String? ?? '未命名设定',
      type: json['type'] as String?,
      content: json['content'] as String?,
      description: json['description'] as String?,
      attributes: attributesMap,
      imageUrl: json['imageUrl'] as String?,
      relationships: relationships,
      sceneIds: sceneIdsList,
      priority: json['priority'] as int?,
      status: json['status'] as String?,
      generatedBy: json['generatedBy'] as String?,
      tags: tagsList,
      vector: vectorList,
      createdAt: json['createdAt'] != null ? DateTime.tryParse(json['createdAt'].toString()) : null,
      updatedAt: json['updatedAt'] != null ? DateTime.tryParse(json['updatedAt'].toString()) : null,
      isAiSuggestion: json['isAiSuggestion'] as bool? ?? false,
      metadata: metadataMap,
      parentId: json['parentId'] as String?,
      childrenIds: childrenIdsList,
      nameAliasTracking: NameAliasTracking.fromValue(json['nameAliasTracking'] as String?),
      aiContextTracking: AIContextTracking.fromValue(json['aiContextTracking'] as String?),
      referenceUpdatePolicy: SettingReferenceUpdate.fromValue(json['referenceUpdatePolicy'] as String?),
    );
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = <String, dynamic>{};
    if (id != null) data['id'] = id;
    if (novelId != null) data['novelId'] = novelId;
    if (userId != null) data['userId'] = userId;
    data['name'] = name;
    if (type != null) data['type'] = type;
    if (content != null) data['content'] = content;
    if (description != null) data['description'] = description;
    if (attributes != null) data['attributes'] = attributes;
    if (imageUrl != null) data['imageUrl'] = imageUrl;
    if (relationships != null) {
      data['relationships'] = relationships!.map((e) => e.toJson()).toList();
    }
    if (sceneIds != null) data['sceneIds'] = sceneIds;
    if (priority != null) data['priority'] = priority;
    if (generatedBy != null) data['generatedBy'] = generatedBy;
    if (tags != null) data['tags'] = tags;
    if (status != null) data['status'] = status;
    if (vector != null) data['vector'] = vector;
    if (createdAt != null) data['createdAt'] = createdAt!.toIso8601String();
    if (updatedAt != null) data['updatedAt'] = updatedAt!.toIso8601String();
    data['isAiSuggestion'] = isAiSuggestion;
    if (metadata != null) data['metadata'] = metadata;
    if (parentId != null) data['parentId'] = parentId;
    if (childrenIds != null) data['childrenIds'] = childrenIds;
    data['nameAliasTracking'] = nameAliasTracking.value;
    data['aiContextTracking'] = aiContextTracking.value;
    data['referenceUpdatePolicy'] = referenceUpdatePolicy.value;
    return data;
  }

  NovelSettingItem copyWith({
    String? id,
    String? novelId,
    String? userId,
    String? name,
    String? type,
    String? content,
    String? description,
    Map<String, String>? attributes,
    String? imageUrl,
    List<SettingRelationship>? relationships,
    List<String>? sceneIds,
    int? priority,
    String? generatedBy,
    List<String>? tags,
    String? status,
    List<double>? vector,
    DateTime? createdAt,
    DateTime? updatedAt,
    bool? isAiSuggestion,
    Map<String, dynamic>? metadata,
    String? parentId,
    List<String>? childrenIds,
    NameAliasTracking? nameAliasTracking,
    AIContextTracking? aiContextTracking,
    SettingReferenceUpdate? referenceUpdatePolicy,
  }) {
    return NovelSettingItem(
      id: id ?? this.id,
      novelId: novelId ?? this.novelId,
      userId: userId ?? this.userId,
      name: name ?? this.name,
      type: type ?? this.type,
      content: content ?? this.content,
      description: description ?? this.description,
      attributes: attributes ?? this.attributes,
      imageUrl: imageUrl ?? this.imageUrl,
      relationships: relationships ?? this.relationships,
      sceneIds: sceneIds ?? this.sceneIds,
      priority: priority ?? this.priority,
      generatedBy: generatedBy ?? this.generatedBy,
      tags: tags ?? this.tags,
      status: status ?? this.status,
      vector: vector ?? this.vector,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      isAiSuggestion: isAiSuggestion ?? this.isAiSuggestion,
      metadata: metadata ?? this.metadata,
      parentId: parentId ?? this.parentId,
      childrenIds: childrenIds ?? this.childrenIds,
      nameAliasTracking: nameAliasTracking ?? this.nameAliasTracking,
      aiContextTracking: aiContextTracking ?? this.aiContextTracking,
      referenceUpdatePolicy: referenceUpdatePolicy ?? this.referenceUpdatePolicy,
    );
  }

  @override
  List<Object?> get props => [
    id, novelId, userId, name, type, content, description, attributes, 
    imageUrl, relationships, sceneIds, priority, generatedBy, tags, status, 
    vector, createdAt, updatedAt, isAiSuggestion, metadata, parentId, childrenIds,
    nameAliasTracking, aiContextTracking, referenceUpdatePolicy
  ];

  @override
  String toString() {
    return jsonEncode(toJson());
  }
}

/// 设定关系模型
class SettingRelationship extends Equatable {
  final String targetItemId;
  final SettingRelationshipType type;
  final String? description;
  final int? strength;
  final String? direction;
  final DateTime? createdAt;
  final Map<String, dynamic>? attributes;

  const SettingRelationship({
    required this.targetItemId,
    required this.type,
    this.description,
    this.strength,
    this.direction,
    this.createdAt,
    this.attributes,
  });

  factory SettingRelationship.fromJson(Map<String, dynamic> json) {
    return SettingRelationship(
      targetItemId: json['targetItemId'] as String,
      type: SettingRelationshipType.fromValue(json['type'] as String),
      description: json['description'] as String?,
      strength: json['strength'] as int?,
      direction: json['direction'] as String?,
      createdAt: json['createdAt'] != null ? DateTime.tryParse(json['createdAt'].toString()) : null,
      attributes: json['attributes'] != null ? Map<String, dynamic>.from(json['attributes']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = <String, dynamic>{};
    data['targetItemId'] = targetItemId;
    data['type'] = type.value;
    if (description != null) data['description'] = description;
    if (strength != null) data['strength'] = strength;
    if (direction != null) data['direction'] = direction;
    if (createdAt != null) data['createdAt'] = createdAt!.toIso8601String();
    if (attributes != null) data['attributes'] = attributes;
    return data;
  }

  @override
  List<Object?> get props => [targetItemId, type, description, strength, direction, createdAt, attributes];
} 